tests/test_wheel.py

```diff
@@ -249,3 +250,22 @@ def test_editable(
         assert venv.python('-c', 'import plat; print(plat.foo())').strip() == 'something else'
     finally:
         plat.write_text(plat_text)
+
+def test_editable_broken_non_existent_build_dir(
```

Comment:
This PR looks like a useful robustness improvement. It'd be good to add a more explicit comment here about a few of the ways that this can happen (e.g. user does `rm -rf .mesonpy`) and that what needs to happen then is a full rebuild.

Reply:
Unfortunately, I don't have a way to consistently reproduce this. Sometimes, I can still get it to work.

Also, this method is not foolproof. If I delete all the meson folders(meson is somehow resilient to only one folder being deleted) in the build dir, by hand, then we can still fail to regenerate the builddir, even with build.ninja present.

(We probably don't care about that case, though, since I don't see how the build dir could get corrupted like that).

Reply:
Maybe @eli-schwartz has a better idea of what to check for.

Reply:
Meson internally checks to see if the file `meson-private/coredata.dat` exists inside the build directory.

- There are too many reasons an empty directory might exist (including several ways that all files, but no directories, can be deleted by forces outside of Meson).
- The build.ninja isn't suitable, since that doesn't tell us whether it was Meson that created it.
- Internally, we decided to use coredata.dat as the "existence check", so correspondingly, we also arrange to have this file be created as late as we can, or deleted if something went wrong at the last minute.

Reply:
Awesome, this is exactly what I've been looking for.

Also kind of explains, why meson-python fails sometimes when the setup process fails.